### PR TITLE
Feature/orca 420 Research paging of large data returns for code.

### PR DESCRIPTION
**Pull request by AndrewDorn**
## Summary of Changes

Research paging of large data returns for code.

Addresses [ORCA-420: Research paging of large data returns for code.](https://bugs.earthdata.nasa.gov/browse/ORCA-420)

## Changes

* Added page for postgres tips
* Added tips on paging

## Type of change

- [x] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the necessary documentation (remove those that do not apply)
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes
    - [x] Automated tests passing (if not fork)
- [x] Any dependent changes have been merged and published in downstream modules
- [x] My code has passed security scanning
    - [x] Snyk
    - [x] git-secrets

### CODE REVIEW

**Comment by AndrewDorn** [comment]
I believe the implementation of this will be fairly straigtforward with the current plan, and am personally comfortable with building it into existing graphql tickets.

**Review by rizbihassan — CHANGES_REQUESTED** [review]
lets create cards required to implement these tasks such as updating lambdas, etc.

**Comment by AndrewDorn** [comment]
> lets create cards or refer to existing cards (if any) required to implement these tasks such as updating lambdas, etc.

I would vote against modifying existing lambdas. I view this as something that will be created in Graphql for new/migrated uses.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+```
+
+Adjust code as needed to match your table.
+
+### Page Retrieval
+The initial impulse for retrieving a page may be to `SKIP` the number of entries from previous pages.
+However, this requires the query to process all of the previous entries, and is therefore non-performant on large datasets.
+
+A better option is to apply a `WHERE` clause that picks up after the previous query.
+For example, following from a page where the final entry was `key0: 50, key1: 98, key2: 60`, retrieving the next page of size `100` would look like:
```
Comment (bhazuka) [inline website/docs/developer/development-guide/code/postgres-tips.md:32]: Reword this slightly so it is clear that we are using these values for all the examples below. Maybe like the following

```
In the examples below we are going to assume the row we will paging from has the following unique values of `key0: 50, key1: 98, key2: 60`. In all instances we will be limiting the result to `100` records.
```

You may also want break this into 2 sub sections like forward paging and reverse or backward paging below to better organize paging thoughts.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+```
+SELECT * FROM orca.performance_check_multiple
+WHERE key0 >= 50 AND (key0 > 50 OR (key1 >= 98 AND (key1 > 98 OR (key2 > 60))))
+ORDER BY key0 ASC, key1 ASC, key2 ASC
+LIMIT 100
+```
+Note that the order of the keys in the `WHERE` and `ORDER BY` clauses must be identical in this example.
+The filter should eliminate all of the previously retrieved rows.
+
+In the event of a descending order, flip the signs and order:
```
Comment (bhazuka) [inline website/docs/developer/development-guide/code/postgres-tips.md:42]: I feel this is a little misleading and could be clarified. Possibly use subsection maybe something like this:

```
#### Reverse Paging

In the event the user requests the previous page, a reverse cursor is needed.
Based on the example table and data, to create reverse looking cursor flip the
comparison signs and the direction of the ordering as seen in the  query 
below.
```

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ORDER BY key0 DESC, key1 DESC, key2 DESC
+LIMIT 100
+```
+
+In the event the user requests the previous page, flip the signs and order, then reverse the order of results.
+```
+SELECT * FROM (
+	SELECT * FROM orca.performance_check_multiple
+	WHERE key0 <= 50 AND (key0 < 50 OR (key1 <= 98 AND (key1 < 98 OR (key2 < 60))))
+	ORDER BY key0 ASC, key1 ASC, key2 ASC
```
Comment (bhazuka) [inline website/docs/developer/development-guide/code/postgres-tips.md:55]: I believe this should be `ORDER BY key0 DESC, key1 DESC, key2 DESC`

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+In the event of a descending order, flip the signs and order:
+```
+SELECT * FROM orca.performance_check_multiple
+WHERE key0 <= 50 AND (key0 < 50 OR (key1 <= 98 AND (key1 < 98 OR (key2 < 60))))
+ORDER BY key0 DESC, key1 DESC, key2 DESC
+LIMIT 100
+```
+
+In the event the user requests the previous page, flip the signs and order, then reverse the order of results.
```
Comment (bhazuka) [inline website/docs/developer/development-guide/code/postgres-tips.md:50]: Possibly reword this to something like:

```
In the results from our reverse cursor above, the largest record 
`key0: 50, key1: 98, key2: 59` appears first and the smallest record
in the set  `key0: 50, key1: 97, key2: 60` appears last. To change this
behavior and maintain the same ordering as the forward cursor, wrap
the query with and additional ORDER BY to change the record order
as seen below.
```

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+SELECT * FROM (
+	SELECT * FROM orca.performance_check_multiple
+	WHERE key0 <= 50 AND (key0 < 50 OR (key1 <= 98 AND (key1 < 98 OR (key2 < 60))))
+	ORDER BY key0 ASC, key1 ASC, key2 ASC
+	LIMIT 100
+) results
+ORDER BY key0 ASC, key1 ASC, key2 ASC
+```
+
+# todo: bidirectional
```
Comment (bhazuka) [inline website/docs/developer/development-guide/code/postgres-tips.md:61]: I feel you can remove this TODO as you describe a forward and reverse cursor and how to perform forward paging or reverse paging.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+### Cursors
+`decode_cursor` and `encode_cursor` in the graphql project can be used to encode and decode arbitrary dictionaries for use in future queries.
+In our example table, this would consist of `key0`, `key1`, and `key2` to be able to enforce paging uniqueness.
+These cursors point to individual elements.
+At minimum, users should be given a `start_cursor` and an `end_cursor` for any page they retrieve.
+The customer can then make a request with `order: 'Ascending', cursor: {their end_cursor}, direction: 'Next'` to get the page past the end_cursor,
```
Comment (bhazuka) [inline website/docs/developer/development-guide/code/postgres-tips.md:68]: The `order: Ascending` confuses me here in both instances you refer to it. I guess I would expect the customer to provide something like:
`cursor: {start or end cursor}, direction: {next or previous}, limit: int`. Thoughts?

**Review by bhazuka — CHANGES_REQUESTED** [review]
Lots of good content. Added some thoughts and ideas for clarity, better readability and reproducibility.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+---
+
+## Paging
+It will not always be feasible/desired to return all results at once.
+In this case, results should be returned one 'page' at a time, with cursors included to allow customers to easily request the next page.
+
+### Sample Table
+Code in the following sections will be based off of the following table:
+```
+CREATE TABLE "orca"."test_table"  (
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:14]: Whoops. Names were left-over from a different test. Fixed. Added dummy data, though I kept it non-random.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 8]
```diff
@@ -0,0 +1,69 @@
...
+Adjust code as needed to match your table.
+
+### Page Retrieval
+The initial impulse for retrieving a page may be to `SKIP` the number of entries from previous pages.
+However, this requires the query to process all of the previous entries, and is therefore non-performant on large datasets.
+
+A better option is to apply a `WHERE` clause that picks up after the previous query.
+For example, following from a page where the final entry was `key0: 50, key1: 98, key2: 60`, retrieving the next page of size `100` would look like:
+```
+SELECT * FROM orca.performance_check_multiple
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:34]: Added. Made a couple tweaks to preserve parenthesis alignment.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+WHERE key0 >= 50 AND (key0 > 50 OR (key1 >= 98 AND (key1 > 98 OR (key2 > 60))))
+ORDER BY key0 ASC, key1 ASC, key2 ASC
+LIMIT 100
+```
+Note that the order of the keys in the `WHERE` and `ORDER BY` clauses must be identical in this example.
+The filter should eliminate all of the previously retrieved rows.
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:40]: Added note regarding ascending order. Admittedly, not sure I understand this comment.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 10]
```diff
@@ -0,0 +1,69 @@
...
+
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:41]: Added with minor tweaks.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 11]
```diff
@@ -0,0 +1,69 @@
...
+GO
+ALTER TABLE "orca"."test_table"
+	ADD CONSTRAINT "test_uniqueness"
+	UNIQUE ("key0", "key1", "key2")
+```
+
+Adjust code as needed to match your table.
+
+### Page Retrieval
+The initial impulse for retrieving a page may be to `SKIP` the number of entries from previous pages.
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:28]: Sounds good. Added, and tweaked next section to reference the example.

File: website/docs/developer/development-guide/code/postgres-tips.md [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+However, this requires the query to process all of the previous entries, and is therefore non-performant on large datasets.
+
+A better option is to apply a `WHERE` clause that picks up after the previous query.
```
[the remark this answers is not in the document]
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:31]: True. Added.

File: website/docs/developer/development-guide/code/postgres-tips.md (hunk 1 above)
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:32]: Added line regarding all examples.

File: website/docs/developer/development-guide/code/postgres-tips.md (hunk 2 above)
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:42]: As cursors point to individual rows, they have no direction. The cursor consists only of `key0: 50, key1: 98, key2: 60`

File: website/docs/developer/development-guide/code/postgres-tips.md (hunk 3 above)
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:55]: Good catch. Fixed.

File: website/docs/developer/development-guide/code/postgres-tips.md (hunk 4 above)
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:50]: Added a few words to indicate that this is carrying on from the default rather than the `Descending Order` section.

File: website/docs/developer/development-guide/code/postgres-tips.md (hunk 5 above)
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:61]: Whoops. Forgot I left that there. Removed.

File: website/docs/developer/development-guide/code/postgres-tips.md (hunk 6 above)
Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:68]: I think we should keep control of limit ourselves to avoid users beeing cheeky and giving us 999999999999999999 to avoid paging.

Comment (AndrewDorn) [reply website/docs/developer/development-guide/code/postgres-tips.md:68]: I recall from our talks that users requesting the end of a set of records was desired. In your example, would they do that by sending `cursor: None, direction: 'previous'`? They'd have to be okay with reordering things themselves, since their results would look like [9, 10], [7, 8], [5, 6]...

**Review by bhazuka — APPROVED** [review]

File: CHANGELOG.md [hunk 13]
```diff
@@ -92,7 +92,7 @@ and includes an additional section for migration notes.
 
 ## [6.0.0]
 ### Changed
-- *ORCA-290* Renamed `excludeFileTypes`, `orcaDefaultBucketOverride` and `orcaDefaultStorageClassOverride` to `excludedFileExtensions`, `defaultBucketOverride` and  `defaultStorageClassOverride` respectively. In addition, ORCA configuration variables `excludedFileExtensions`, `defaultBucketOverride` and `defaultStorageClassOverride` are now under `collection.meta.orca`.
+- *ORCA-290* Renamed `excludeFileTypes`, `orcaDefaultBucketOverride`, `orcaDefaultRecoveryTypeOverride`, and `orcaDefaultStorageClassOverride` to `excludedFileExtensions`, `defaultBucketOverride` `defaultRecoveryTypeOverride`, and  `defaultStorageClassOverride` respectively. In addition, ORCA configuration variables `excludedFileExtensions`, `defaultBucketOverride`, `defaultRecoveryTypeOverride`, and `defaultStorageClassOverride` are now under `collection.meta.orca`.
```
Comment (rizbihassan) [inline CHANGELOG.md:95]: comma missing after  `defaultBucketOverride`

**Review by rizbihassan — APPROVED** [review]
one small fix

File: CHANGELOG.md (hunk 13 above)
Comment (AndrewDorn) [reply CHANGELOG.md:95]: Fixed
